subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```
@@ -0,0 +1,805 @@
+			match &= (str && (strcmp(str, msg_type) == 0));
+		}
+
+		return match ? 0 : -ENOMSG;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_str_get(const struct nrf_cloud_obj *const obj, const char *const key,
+	char **str)
+{
+	if (!obj || !key || !str) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		return get_string_from_obj(obj->json, key, str);
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_msg_init(struct nrf_cloud_obj *const obj, const char *const app_id,
+	const char *const msg_type)
+{
+	if (!obj || !app_id) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		if (obj->json) {
+			return -ENOTEMPTY;
+		}
+
+		int err;
+		cJSON *new_obj = cJSON_CreateObject();
+
+		err = (cJSON_AddStringToObjectCS(new_obj, NRF_CLOUD_JSON_APPID_KEY,
+						 app_id) ? 0 : -ENOMEM);
+
+		if (!err && msg_type) {
+			err = (cJSON_AddStringToObjectCS(new_obj,
+							 NRF_CLOUD_JSON_MSG_TYPE_KEY,
+							 msg_type) ? 0 : -ENOMEM);
+		}
+
+		if (err) {
+			cJSON_Delete(new_obj);
+			return -ENOMEM;
+		}
+
+		obj->json = new_obj;
+		return 0;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_init(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		if (obj->json) {
+			return -ENOTEMPTY;
+		}
+
+		obj->json = cJSON_CreateObject();
+		return (obj->json ? 0 : -ENOMEM);
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_reset(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		obj->json = NULL;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		return -ENOTSUP;
+	}
```

I wonder if we should instead raise an error if `obj->json` is not already NULL -- after all, that would suggest we are about to create a memory leak, wouldn't it?

Maybe we shouldn't support `nrf_cloud_obj_reset` at all? Sure, perhaps the user is doing something like taking ownership of the object and then discarding the `nrf_cloud_obj` struct, except they want to re-use it for another decode, but is that really a pattern we want to support?

Maybe there is a use case here that I'm missing, though